src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/DiagnosticDescriptorExtensions.cs

```diff
@@ -97,14 +136,33 @@ public static bool IsDefinedInEditorConfig(this DiagnosticDescriptor descriptor,
             return false;
         }
 
-        public static ReportDiagnostic GetEffectiveSeverity(this DiagnosticDescriptor descriptor, AnalyzerConfigOptions analyzerConfigOptions)
+        public static ReportDiagnostic GetEffectiveSeverity(
```

Comment:
💭 It's really hard to tell which method in this file is the "right one" to use. I would expect there to be one fully-implemented method that gives the right answer, and the rest of the methods delegate to it somehow.

Reply:
> It's really hard to tell which method in this file is the "right one" to use.

That is indeed true. However, the issue is the way these options bags are represented so differently across different layers and components. Some have it as string based dictionary, some as an opaque type that allow `TryGetValue(key, out value)`. Some of these types require providing a syntax tree for the path for which to get an option and some of them just allow querying based on the key. I haven't been able to find a way to fully unify them, so tried to minimize the duplication in this PR by just adding an optional parameter to an existing method instead of writing another fork of it based off syntax tree options provider + syntax tree.